File: code_eval/cpp/moderate/challenge_12.cpp
// Created by cschalk on 3/22/16.
// Challenge 12 : First non duplicate letter
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

template<typename T, typename U>
inline bool contains(T const &t, U const &u) {
    auto result = std::find(t.begin(), t.end(), u);
    return result != t.end();
}

template<typename T, typename U>
inline void remove(T &t, U const &u) {
    t.erase(std::remove(t.begin(), t.end(), u), t.end());
}

int main(int argc, char ** args) {
    std::ifstream input {args[1]};
    std::string line;

    std::vector<char> foundDuplicates;
    std::vector<char> foundLetters;
    while (std::getline(input, line)) {
        foundDuplicates.clear();
        foundLetters.clear();
        for (std::string::size_type i = 0 ; i < line.length() ; ++i ) {
            if (!contains(foundDuplicates, line[i])) {
                if (!contains(foundLetters, line[i])) {
                    foundLetters.push_back(line[i]);
                } else {
                    remove(foundLetters, line[i]);
                    foundDuplicates.push_back(line[i]);
                }
            }
        }
        if (!foundLetters.empty()) {
            std::cout << foundLetters.front() << std::endl;
        }
    }

    return 0;

}

File: code_eval/java/moderate/Challenge12.java
package moderate;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;

/**
 * Challenge 12 : First non-duplicate letter
 * Created by cschalk on 3/21/16.
 */
public class Challenge12 {
    public static void main (String[] args) throws Exception {
        File file = new File(args[0]);
        BufferedReader buffer = new BufferedReader(new FileReader(file));
        String line;
        while ((line = buffer.readLine()) != null) {
            runChallenge(line);
        }
    }

    private static void runChallenge(String input) {
        ArrayList<String> foundLetters = new ArrayList<>();
        ArrayList<String> foundDuplicates = new ArrayList<>();
        for (char c : input.toCharArray()) {
            String work = "" + c;
            if (!foundDuplicates.contains(work)) {
                if (!foundLetters.contains(work)) {
                    foundLetters.add(work);
                } else {
                    foundLetters.remove(work);
                    foundDuplicates.add(work);
                }
            }
        }
        if (foundLetters.size() > 0)
            System.out.println(foundLetters.get(0));
    }
}
